core/contracts/Registry.sol

```diff
@@ -57,58 +61,103 @@ contract Registry is RegistryInterface, MultiRole {
+        return addressToDerivatives[derivative].valid == DerivativeValidity.Valid;
     }
 
-    function getRegisteredDerivatives(address party) external view returns (address[] memory derivatives) {
-        // This is not ideal - we must statically allocate memory arrays. To be safe, we make a temporary array as long
-        // as registeredDerivatives. We populate it with any derivatives that involve the provided party. Then, we copy
-        // the array over to the return array, which is allocated using the correct size. Note: this is done by double
-        // copying each value rather than storing some referential info (like indices) in memory to reduce the number
-        // of storage reads. This is because storage reads are far more expensive than extra memory space (~100:1).
-        address[] memory tmpDerivativeArray = new address[](registeredDerivatives.length);
-        uint outputIndex = 0;
-        for (uint i = 0; i < registeredDerivatives.length; i = i.add(1)) {
-            address derivative = registeredDerivatives[i];
-            if (derivativesToParties[derivative].parties[party]) {
-                // Copy selected derivative to the temporary array.
-                tmpDerivativeArray[outputIndex] = derivative;
-                outputIndex = outputIndex.add(1);
-            }
-        }
+    function getRegisteredDerivatives(address party) external view returns (address[] memory) {
+        return partyMembersToDerivatives[party].derivatives;
+    }
+
+    function addPartyToDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Can add to valid derivative"
+        );
+        require(!isPartyMemberOfDerivative(party, derivativeAddress), "Can only register a party once");
+
+        // Push the derivative and store the index.
+        uint derivativeIndex = partyMembersToDerivatives[party].derivatives.push(derivativeAddress);
+        partyMembersToDerivatives[party].derivativeIndex[derivativeAddress] = derivativeIndex - 1;
 
-        // Copy the temp array to the return array that is set to the correct size.
-        derivatives = new address[](outputIndex);
-        for (uint j = 0; j < outputIndex; j = j.add(1)) {
-            derivatives[j] = tmpDerivativeArray[j];
+        emit PartyMemberAdded(derivativeAddress, party);
+    }
+
+    function removePartyFromDerivative(address party) external {
+        address derivativeAddress = msg.sender;
+
+        PartyMember storage partyMember = partyMembersToDerivatives[party];
+
+        require(
+            addressToDerivatives[derivativeAddress].valid == DerivativeValidity.Valid,
+            "Remove only from valid derivative"
+        );
+        require(isPartyMemberOfDerivative(party, derivativeAddress), "Can only register a party once");
```

Comment:
Is this error message accurate? Shouldn't it be something like "Can only remove an existing party member"